File: domain/src/main/kotlin/com/pricelens/domain/trust/Reputation.kt
package com.pricelens.domain.trust

import kotlin.math.sqrt

/**
 * Implements the Beta-Bernoulli reputation model with Lower Confidence Bound (H-08).
 * Ensures reputation is slow to earn and fast to lose.
 */
data class Reputation(
    val agreements: Double = 0.0,
    val disagreements: Double = 0.0
) {
    /**
     * Calculates the reputation score as the Lower Confidence Bound (P10) of the Beta distribution.
     */
    fun score(): Float {
        val a = agreements + 1.0
        val b = disagreements + 1.0
        val sum = a + b
        
        val mean = a / sum
        val variance = (a * b) / (sum * sum * (sum + 1.0))
        val stdDev = sqrt(variance)
        
        // z = 1.28 for P10 (10th percentile)
        val lcb = mean - 1.28 * stdDev
        
        return lcb.toFloat().coerceIn(0.0f, 1.0f)
    }

    /**
     * Updates reputation based on a new outcome.
     * Disagreements carry 2.5x the weight of agreements.
     */
    fun update(isAgreement: Boolean, weight: Float = 1.0f): Reputation {
        return if (isAgreement) {
            copy(agreements = agreements + weight)
        } else {
            // "Fast to lose": disagreements weighted 2.5x
            copy(disagreements = disagreements + weight * 2.5)
        }
    }
}

File: backend/app/services/reputation.py
import numpy as np
from scipy.stats import beta
from typing import Tuple

class ReputationEngine:
    """
    Implements Beta-Bernoulli reputation scoring with Lower Confidence Bound.
    Reputation is category-specific and slow to earn, fast to lose.
    """

    # Weight of a single disagreement vs agreement
    DISAGREEMENT_PENALTY = 2.5

    def update_scores(self, alpha: float, beta_val: float, agreed: bool, weight: float) -> Tuple[float, float]:
        """
        Updates alpha/beta parameters based on a new observation outcome.
        """
        if agreed:
            new_alpha = alpha + weight
            new_beta = beta_val
        else:
            new_alpha = alpha
            new_beta = beta_val + (weight * self.DISAGREEMENT_PENALTY)

        return new_alpha, new_beta

    def calculate_lcb(self, alpha_val: float, beta_val: float, confidence: float = 0.90) -> float:
        """
        Calculates the P10 (10th percentile) of the Beta distribution.
        This ensures users with few observations are not over-trusted.
        """
        # Prior is Beta(1, 1) - uniform
        a = alpha_val + 1.0
        b = beta_val + 1.0

        # ppf is the percent point function (inverse of cdf)
        # We want the 10th percentile (P10)
        return float(beta.ppf(1.0 - confidence, a, b))
